src/dict.c

```diff
@@ -675,23 +694,36 @@ void dictRelease(dict *d)
 
 dictEntry *dictFind(dict *d, const void *key)
 {
-    dictEntry *he;
-    uint64_t h, idx, table;
+    dictEntry *he = NULL;
+    uint64_t h, idx;
 
     if (dictSize(d) == 0) return NULL; /* dict is empty */
-    if (dictIsRehashing(d)) _dictRehashStep(d);
+
     h = dictHashKey(d, key);
-    for (table = 0; table <= 1; table++) {
+    idx = h & DICTHT_SIZE_MASK(d->ht_size_exp[0]);
+    if ((long)idx >= d->rehashidx) {
+        he = d->ht_table[0][idx];
+    }
+
+    if (dictIsRehashing(d)) {
+        if ((long)idx >= d->rehashidx && he) {
+            dictBucketRehash(d, h);
+        /* If the 'he' is not in ht0, we perform a random bucket rehash. */
+        } else {
+            assert(!he);
+            _dictRehashStep(d);
+        }
+        /* If rehashing is still ongoing, we lookup in table 1, otherwise in table 0. */
+        int table = dictIsRehashing(d) ? 1 : 0;
         idx = h & DICTHT_SIZE_MASK(d->ht_size_exp[table]);
-        if (table == 0 && (long)idx < d->rehashidx) continue; 
         he = d->ht_table[table][idx];
-        while(he) {
-            void *he_key = dictGetKey(he);
-            if (key == he_key || dictCompareKeys(d, key, he_key))
-                return he;
-            he = dictGetNext(he);
-        }
-        if (!dictIsRehashing(d)) return NULL;
+    }
+
+    while(he) {
```

Comment:
let's add a title for this block.
e.g. `we have the right table and bucket, find the right key.